x-pack/plugins/security_solution/server/lib/detection_engine/notifications/schedule_notification_response_actions.ts

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { map, uniq } from 'lodash';
+import type { SetupPlugins } from '../../../plugin_contract';
+
+interface OsqueryQuery {
+  id: string;
+  query: string;
+  ecs_mapping: Record<string, Record<'field', string>>;
+  version: string;
+  platform: string;
+}
+
+interface OsqueryResponseAction {
+  actionTypeId: '.osquery';
+  params: {
+    id: string;
+    queries: OsqueryQuery[];
+    savedQueryId: string;
+    query: string;
+    packId: string;
+    ecs_mapping?: Record<string, { field?: string; value?: string }>;
+  };
+}
+
+export interface ResponseAction {
+  actionTypeId: string;
+  params: { [p: string]: unknown };
+}
+
+interface ScheduleNotificationActions {
+  signals: unknown[];
+  responseActions: ResponseAction[];
+}
+
+interface IAlert {
+  agent: {
+    id: string;
+  };
+}
+
+const isOsqueryAction = (action: ResponseAction): action is OsqueryResponseAction => {
+  return action.actionTypeId === '.osquery';
+};
+
+export const scheduleNotificationResponseActions = (
+  { signals, responseActions }: ScheduleNotificationActions,
+  osqueryCreateAction?: SetupPlugins['osquery']['osqueryCreateAction']
+) => {
+  const filteredAlerts = (signals as IAlert[]).filter((alert) => alert.agent?.id);
+  const agentIds = uniq(filteredAlerts.map((alert: IAlert) => alert.agent?.id));
+  const alertIds = map(filteredAlerts, '_id');
```

Comment:
This cast `signals as IAlert[]` can be dangerous. @marshallmain can `result.createdSignals` passed to this function be a bit more specific than `unknown[]`?

Reply:
I think we could update the types so the `createdSignals` are typed more strictly. Right now the problem is that `searchAfterAndBulkCreate` takes the created alerts, that are statically typed, and returns them as `unknown` instead. It should be able to return them as the same type that `bulkCreate` returns. However, the alerts don't include `agent` fields unless they're copied from the source docs, so `agent` fields aren't in the static type and this runtime check is still necessary.